### PR TITLE
Don't propagate cancel signal to the Prometheus rules manager context

### PR DESCRIPTION
**Pull request by rapphil**
**What this PR does**:

* Allow rules that are being evaluated to successfully complete/timeout when rulers are shutdown. Before this PR, we were canceling the context right away, wasting all the work performed by the queries.

This is done by removing the cancel signal of the context passed to the Prometheus rules manager. Instead we are going to rely on the shutdown mechanism of the rules manager and timeout of the queries executed by the rules.

In the unit tests I tried to avoid adding sleeps and instead used channels to coordinate the execution of the queries to simulate the problem. 

**Checklist**
- [X] Tests updated
- [ ]  Documentation added - N.A.
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ruler/compat.go [hunk 1]
```diff
@@ -341,11 +341,15 @@ func DefaultTenantManagerFactory(cfg Config, p Pusher, q storage.Queryable, engi
 			queryFunc = metricsQueryFunc
 		}
 
+		// We let the Prometheus rules manager control the context so that there is a chance
+		// for graceful shutdown of rules that are still in execution even in case the cortex context is canceled.
+		prometheusContext := user.InjectOrgID(context.WithoutCancel(ctx), userID)
```
Comment (alanprot) [inline pkg/ruler/compat.go:346]: til! :D

Comment (rajagopalanand) [reply pkg/ruler/compat.go:346]: TIL

Comment (rapphil) [reply pkg/ruler/compat.go:346]: this was added in go1.21.0

**Comment by alanprot** [comment]
LGTM!

**Review by alanprot — APPROVED** [review]

**Review by yeya24 — APPROVED** [review]

**Pull request merged**